libpoolprotocols/stratum/EthStratumClient.cpp

```diff
@@ -1,1459 +1,1489 @@
+{
+    dev::setThreadName("stratum");
+
+    // Set status completion
+    m_conntimer.cancel();
+    m_connecting.store(false, std::memory_order_relaxed);
+
+    // Timeout has run before or we got error
+    if (ec || !m_socket->is_open())
+    {
+        cwarn << ("Error  " + toString(m_endpoint) + " [ " + ( ec ? ec.message() : "Timeout") + " ]");
+
+        // We need to close the socket used in the previous connection attempt
+        // before starting a new one.
+        // In case of error, in fact, boost does not close the socket
+        // If socket is not opened it means we got timed out
+        if (m_socket->is_open())
+            m_socket->close();
+
+        // Discard this endpoint and try the next available.
+        // Eventually is start_connect which will check for an
+        // empty list.
+        m_endpoints.pop();
+        m_canconnect.store(false, std::memory_order_relaxed);
+        m_io_service.post(m_io_strand.wrap(boost::bind(&EthStratumClient::start_connect, this)));
+
+        return;
+    }
+
+    // We got a socket connection established
+    m_canconnect.store(true, std::memory_order_relaxed);
+    if (g_logVerbosity >=6 )
+        cnote << "Socket connected to " << ActiveEndPoint();
+
+    if (m_conn->SecLevel() != SecureLevel::NONE)
+    {
+        boost::system::error_code hec;
+        m_securesocket->lowest_layer().set_option(boost::asio::socket_base::keep_alive(true));
+        m_securesocket->lowest_layer().set_option(tcp::no_delay(true));
+
+        m_securesocket->handshake(boost::asio::ssl::stream_base::client, hec);
+
+        if (hec)
+        {
+            cwarn << "SSL/TLS Handshake failed: " << hec.message();
+            if (hec.value() == 337047686)
+            {  // certificate verification failed
+                cwarn << "This can have multiple reasons:";
+                cwarn << "* Root certs are either not installed or not found";
+                cwarn << "* Pool uses a self-signed certificate";
+                cwarn << "Possible fixes:";
+                cwarn << "* Make sure the file '/etc/ssl/certs/ca-certificates.crt' exists and "
+                         "is accessible";
+                cwarn << "* Export the correct path via 'export "
+                         "SSL_CERT_FILE=/etc/ssl/certs/ca-certificates.crt' to the correct "
+                         "file";
+                cwarn << "  On most systems you can install the 'ca-certificates' package";
+                cwarn << "  You can also get the latest file here: "
+                         "https://curl.haxx.se/docs/caextract.html";
+                cwarn << "* Disable certificate verification all-together via command-line "
+                         "option.";
+            }
+
+            // This is a fatal error
+            // No need to try other IPs as the certificate is based on host-name
+            // not ip address. Trying other IPs would end up with the very same error.

```

Comment:
While you are correct the cert is validated against the hostname, its still possible that another ip can present another, valid certificate.

Reply:
In theory this should not happen and would be an error.
When you try to connect to, say, "eu1.ethermine.org:5555" you get from DNS query only A records bound to that host. This implies ALL services on IPs bound to that specific host must share the very same certificate. 

As a side note I am pretty puzzled about how the validation of certificate on ethminer's side may work as nowhere in the handshaking process we say which is the host name.